### PR TITLE
[AIRFLOW-6817] - Move airflow __init__.py imports to sub-packages

### PR DESCRIPTION
**Pull request by msb217**
---
Issue link: [AIRFLOW-6817](https://issues.apache.org/jira/browse/AIRFLOW-6817)

Moved `airflow/__init__.py` imports to respective sub-packages with the exception of the `settings.initialize()` and `integrate_plugins()`

All breeze unit and integration tests are passing locally.

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/__init__.py
Comment (potiuk) [inline airflow/__init__.py:39]: I think we also need to remove plugins eventually - but this is a good step (and it will be a bit more complex as we have to make sure plugins are initialized where needed).

File: airflow/providers/qubole/example_dags/example_qubole_sensor.py
Comment (potiuk) [inline airflow/providers/qubole/example_dags/example_qubole_sensor.py]: Unrelated?

**Review by potiuk — CHANGES_REQUESTED** [review]
Love it! Sorry for slight delay with reviewing
I think you will need to rebase (and there are a few changes coming from @nuclearpinguin with small refactorings which are probably going to be merged faster.

For sure you need to add a note about UPDATING.md. This is a potentially breaking change as a number of DAGs will have to be adjusted so we need a small table describing which imports were removed and how they should be changed.

The process we follow is that in UPDATING.md we describe how to do it manually but closer to the date we will add some migration assistant which will do those modifications automatically for those who would like to do it automatically. But this is a separate step and for now we need a short description on "how to modify your DAGs after this change".

File: airflow/__init__.py
Comment (msb217) [reply airflow/__init__.py:39]: To do this I'm assuming you'd like to see the `integrate_plugins()` functionality decoupled? 

So for example, `integrate_operator_plugins()` would correspond to the `airflow.operators` module. The function `integrate_sensor_plugins()` would correspond to the `airflow.sensors` module and so on for hooks/macros. 

I'm sure it'd be more complex than what I'm describing, but the idea of only initializing plugins for their respective sub-modules is what you'd be going for, correct?

File: airflow/providers/qubole/example_dags/example_qubole_sensor.py
Comment (msb217) [reply airflow/providers/qubole/example_dags/example_qubole_sensor.py]: Yes - sorry I accidentally included this after a rebase. When I squash/rebase this should be resolved

**Comment by msb217** [comment]
@potiuk  Great to hear you like it. I have provided an `Updating.md` memo regarding these changes and will squash/rebase my commits before the merge.

**Review by potiuk — APPROVED** [review]
LGTM. 

But I would love other committers to take a look. It's a huge one and super important and I might have missed something. @ashb @kaxil @nuclearpinguin @feluelle @BasPH @mik-laj ?

**Comment by potiuk** [comment]
Plus you will need to rebase it @msb217 !

File: airflow/__init__.py
Comment (potiuk) [reply airflow/__init__.py:39]: Something like that. I already had a PR for that (but I think I abandon it as there were many changes in the meantime #6696 )

Comment (potiuk) [reply airflow/__init__.py:39]: But no matter what - if we do it, it should be separate PR.

File: tests/jobs/test_local_task_job.py
Comment (turbaszek) [inline tests/jobs/test_local_task_job.py]: Should we import DAG explicitly?

File: tests/operators/test_bash.py
Comment (turbaszek) [inline tests/operators/test_bash.py]:
```suggestion
from airflow.models.dag import DAG
from airflow.exceptions import AirflowException
```

File: tests/operators/test_latest_only_operator.py
Comment (turbaszek) [inline tests/operators/test_latest_only_operator.py]:
```suggestion
from airflow.models.dag import DAG
from airflow import settings
```

File: tests/cli/commands/test_sync_perm_command.py
Comment (turbaszek) [inline tests/cli/commands/test_sync_perm_command.py]: Should we import DAG directly from `airflow.models.dag`?

**Comment by nuclearpinguin** [comment]
I love this change! <3

File: tests/jobs/test_local_task_job.py
Comment (msb217) [reply tests/jobs/test_local_task_job.py]: Yes I think that would be good for consistency.

Would it be worth considering a similar change to `airflow/models/__init__.py`? Removing the implicit imports of models?

The implicit imports of models are used quite frequently

Comment (potiuk) [reply tests/jobs/test_local_task_job.py]: This is a bit tricky, because we are relying on it in "resetdb" case. The models when imported register themselves in sqlalchemy and resetdb uses that information to know which tables to delete. However we already have some exceptions to that - some models are moved out of models package __init__.py and have to be added manually to resetdb. 

I think what would make more sense is to have a separate models/all_models.py which would import all models and this one can be used by resetdb,

But I think we should make a separate PR for that.

Comment (potiuk) [reply tests/jobs/test_local_task_job.py]: I created an issue for that https://issues.apache.org/jira/browse/AIRFLOW-6870

**Review by turbaszek — APPROVED** [review]

**Review by kaxil — APPROVED** [review]

**Comment by potiuk** [comment]
Thanks @msb217 ! First commit and so important for us !

File: tests/jobs/test_local_task_job.py
Comment (turbaszek) [reply tests/jobs/test_local_task_job.py]: > I think what would make more sense is to have a separate models/all_models.py which would import all models and this one can be used by resetdb,

I am happy to do that as I am pylinting core code :D

**Comment by msb217** [comment]
Thanks @potiuk , I am happy to contribute and hope that I can help with more issues in the future!

**Comment by ashb** [comment]
This is going to need a change to the vast majority of dag in existence - our own docs http://airflow.apache.org/docs/stable/tutorial.html say `from airflow import DAG` .

Is there any corner tricks we can do to have that import still work but only import models etc when first accessed?

**Comment by BasPH** [comment]
I like the change but also with @ashb here, it's going to break a lot of workflows which import `from airflow import DAG`. We should at least show deprecation warnings.

@ashb I don't understand your question? I know of the magic `__import__` function for changing import behaviour, but don't know if it can be applied to your question?

**Comment by ashb** [comment]
For python 3 we can possibly do https://www.python.org/dev/peps/pep-0562/ do have a airflow.DAG still work, but issue a depreciation warning. I'll take a look at a warning on Monday

**Comment by potiuk** [comment]
One thing to consider.

I would like to see if we can think slightly differently here.

This is a change that is super easy to add to automated "migration assistant" tool that we will have to do anyway. I think we will have far more issues to deal with in terms of backwards compatibility (for example bash_operator and python_operator have moved).

I think personally that with the 2.0 migration we should simply make people aware that they have to migrate using migration tool (or manually). Otherwise they will have a false sense of having "compatible" dags. I think telling people they will be able to move their dags to 2.0 and they will continue to work as usual, is a bit miscommunication (or even a bit of a lie - we know it's not going to work in a number of cases). 

Migration to 2.0 will be anyway pain for many people but they can deal with one pain at at time - i.e. DAG migration first, then migrate to 2.0 Airflow infrastructure. 

We can make it  easier for people to migrate to 2.0 - if they will have the new "back-ported" providers packages installed - the migration to the "both 1.10 and 2.0" compatible dags might be done automatically even BEFORE migration to 2.0 and our tool should be able to tell how "ready" you are to migrate to 2.0. For now - I believe none of the changes we've done prevents us from having Dags that are both 1.0 and 2.0 compatible (and if we find one - we can still easily fix it in master to keep that possibility).

The "migration assistant" can produce both - 1.10 and 2.0 compatible DAGs (for example imports changed to "airflow.models.DAG" etc. This should be very easy with tools like Bowler : https://pybowler.io/  - we used it successfully to refactor tons of python code (basically all providers code!) and we can easily produce dag-migration assistant that can be used by people.

**Comment by potiuk** [comment]
I think we have the easiest migration path for our users to follow:
```
[1.10] py2 (those still on py2) -> [1.10] py3 -> [1.10] py3 + providers (incremental) -> [1.10] py3 + providers + 2.0-compatible DAGs (incremental) -> [2.0] 
```

This way while migrating the users will have to deal with one

**Comment by ashb** [comment]
I've thought about this a bit more, and I'm coming at this from a different point of view than simple migration - that of user facing API and what we should expose to them.

Also the easier we make the migration the more likely it is to not leave users behind. So if they _can_ run old dags with new Airflow so much the better. Lets take a lesson from Django here, they've thought long and hard about how to do migrations and upgrades. In summary:

If code works in one version without warnings, it MUST work in the next version but can issue deprecation warnings, and can only be removed in the release after that.

Also to me, `from airflow import DAG` is nicer -- why should the DAG authors have to care about the exact internal implementation of where we've chosen to move this class to now?. Airflow and DAGS go hand in hand -- with PEP-562 we can have lazily imported classes but still present a nice interface to users. And plus we can then re-arrange things later under the hood without having to affect our users. After all, what is the harm if the importing everything is solved?

My proposal here:

- Add in a `_getattr_` based lazy import for DAG to airflow/__init__.py module
- Do _NOT_ issue a deprecation warning for this
- Revert the change to all the imports in example dags etc so that those do `from airflow import DAG`.

**Comment by ashb** [comment]
I'm raising the point about what should our API be on the mailing list.

Plus this PR didn't (yet) have much of an impact on import times - https://github.com/apache/airflow/pull/7505 :)

**Comment by potiuk** [comment]
Not yet indeed :).

Let's talk at the devlist indeed. I understand where you are coming from, but I think it's not good idea for the users in the future to switch to that.

While I understand (and concur) we might want to add PEP-562 behavior for backwards compatibility, I don't think it should be default behavior for examples and future users:

- explicit is better than implicit
- IDE support and mypy will go crazy (I believe - I would like to check it when it's done)

I think a lot of our users are not super-hackers and they use IDEs to write the code. If the PEP-562 results with having no autocomplete/verification of proper import, then it's a no-go for me.

**Pull request merged**
